File: backend/schemas/daytona.py
"""
Schemas for Daytona workspace integration
Daytona sandboxes are only available for ROS projects.
"""
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
from enum import Enum


class WorkspaceState(str, Enum):
    CREATING = "creating"
    RUNNING = "running"
    STOPPED = "stopped"
    ERROR = "error"


class DaytonaWorkspaceCreate(BaseModel):
    """Request to create a Daytona workspace"""
    project_id: str
    repository_url: Optional[str] = None
    branch: Optional[str] = "main"
    ide: str = "vscode"  # vscode, cursor, etc.
    env_vars: Dict[str, str] = {}


class DaytonaWorkspaceResponse(BaseModel):
    """Daytona workspace information"""
    workspace_id: str
    project_id: str
    state: WorkspaceState
    url: Optional[str] = None  # IDE access URL
    ssh_url: Optional[str] = None
    created_at: str
    metadata: Dict[str, Any] = {}
    files_synced: int = 0
    sync_status: str = "none"  # none, syncing, synced, partial, failed


class SyncFilesRequest(BaseModel):
    """Request to sync project files to sandbox"""
    project_id: str


class SyncFilesResponse(BaseModel):
    """Result of file sync operation"""
    sandbox_id: str
    files_synced: int
    sync_status: str  # synced, partial, failed
    errors: List[str] = []


class CodeExecutionRequest(BaseModel):
    """Request to execute code in Daytona workspace"""
    workspace_id: str
    code: str
    language: str = "python"
    timeout: int = 30  # seconds
    env_vars: Dict[str, str] = {}


class CodeExecutionResponse(BaseModel):
    """Result of code execution"""
    success: bool
    stdout: str
    stderr: str
    exit_code: int
    execution_time: float
    error: Optional[str] = None


class SandboxFileEntry(BaseModel):
    """A single entry (file or directory) in the sandbox filesystem"""
    path: str   # relative to PROJECT_BASE_DIR
    type: str   # "file" or "dir"


class SandboxFileListResponse(BaseModel):
    """Filesystem entries present in the sandbox"""
    entries: List[SandboxFileEntry]


class SandboxFileContentResponse(BaseModel):
    """Content of a single sandbox file"""
    path: str
    content: str

File: frontend/lib/api/daytona.ts
/**
 * API client for Daytona workspace management
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api';

export enum WorkspaceState {
  CREATING = 'creating',
  RUNNING = 'running',
  STOPPED = 'stopped',
  ERROR = 'error',
}

export interface DaytonaWorkspace {
  workspace_id: string;
  project_id: string;
  state: WorkspaceState;
  url?: string;
  ssh_url?: string;
  created_at: string;
  metadata: Record<string, any>;
  files_synced: number;
  sync_status: string; // "none" | "syncing" | "synced" | "partial" | "failed"
}

export interface SyncFilesResponse {
  sandbox_id: string;
  files_synced: number;
  sync_status: string;
  errors: string[];
}

export interface CodeExecutionRequest {
  workspace_id: string;
  code: string;
  language: string;
  timeout?: number;
  env_vars?: Record<string, string>;
}

export interface CodeExecutionResponse {
  success: boolean;
  stdout: string;
  stderr: string;
  exit_code: number;
  execution_time: number;
  error?: string;
}

export interface SandboxFileEntry {
  path: string;          // relative to project root
  type: 'file' | 'dir';
}

export interface SandboxFileListResponse {
  entries: SandboxFileEntry[];
}

export interface SandboxFileContentResponse {
  path: string;
  content: string;
}

export const daytonaApi = {
  async createWorkspace(projectId: string, repositoryUrl?: string): Promise<DaytonaWorkspace> {
    const response = await fetch(`${API_BASE}/daytona/workspaces`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        project_id: projectId,
        repository_url: repositoryUrl,
        ide: 'vscode',
        env_vars: {},
      }),
    });
    if (!response.ok) throw new Error('Failed to create workspace');
    return response.json();
  },

  async getWorkspace(workspaceId: string): Promise<DaytonaWorkspace> {
    const response = await fetch(`${API_BASE}/daytona/workspaces/${workspaceId}`);
    if (!response.ok) throw new Error('Failed to fetch workspace');
    return response.json();
  },

  async stopWorkspace(workspaceId: string): Promise<void> {
    const response = await fetch(`${API_BASE}/daytona/workspaces/${workspaceId}`, {
      method: 'DELETE',
    });
    if (!response.ok) throw new Error('Failed to stop workspace');
  },

  async syncFiles(workspaceId: string, projectId: string): Promise<SyncFilesResponse> {
    const response = await fetch(`${API_BASE}/daytona/workspaces/${workspaceId}/sync`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ project_id: projectId }),
    });
    if (!response.ok) throw new Error('Failed to sync files');
    return response.json();
  },

  async executeCode(request: CodeExecutionRequest): Promise<CodeExecutionResponse> {
    const response = await fetch(`${API_BASE}/daytona/execute`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error('Failed to execute code');
    return response.json();
  },

  async listFiles(workspaceId: string): Promise<SandboxFileListResponse> {
    const response = await fetch(`${API_BASE}/daytona/workspaces/${workspaceId}/files`);
    if (!response.ok) throw new Error('Failed to list sandbox files');
    return response.json();
  },

  async getFileContent(workspaceId: string, filePath: string): Promise<SandboxFileContentResponse> {
    const encoded = encodeURIComponent(filePath);
    const response = await fetch(
      `${API_BASE}/daytona/workspaces/${workspaceId}/file-content?path=${encoded}`
    );
    if (!response.ok) throw new Error('Failed to fetch file content');
    return response.json();
  },

  async deletePath(workspaceId: string, path: string): Promise<void> {
    const encoded = encodeURIComponent(path);
    const response = await fetch(`${API_BASE}/daytona/workspaces/${workspaceId}/files?path=${encoded}`, {
      method: 'DELETE',
    });
    if (!response.ok) {
      const text = await response.text();
      throw new Error(text || 'Failed to delete sandbox path');
    }
  },
};
